feat: add component wasm module

ricochet opened the pull request.

Commit:
feat: add component wasm module

carlsverre commented on src/main.rs:
```diff
@@ -1,65 +1,55 @@
-            (param i32 i32)
-            (result i32)
-        ))
+#[allow(unused)]
+use component::*;
```
can we use explicit imports here so it's easier to understand the code without an IDE?

carlsverre commented on src/main.rs:
```diff
@@ -1,65 +1,55 @@
+    // There's a few ways we can call the `answer` `Func` value. The easiest
+    // is to statically assert its signature with `typed` (in this case
+    // asserting it takes no arguments and returns one i32) and then call it.
+    let add = instance.get_typed_func::<(i32, i32), i32, _>(&mut store, "consume_add")?;
```
can we call moo as well?

ricochet replied:
I'm working on lolcat + cowsay :cow2:

carlsverre commented on component/src/lib.rs:
```diff
@@ -0,0 +1,13 @@
+
+impl component::Component for Component {
+    fn consume_add(lh: i32, rh: i32) -> i32 {
+        lh + rh
```
this is awesome, but we have lost the round trip back to the host (showing 2-way interface type calling)

np I can add it back in

carlsverre reviewed (approved):
this is awesome

I assume rust nightly is due to nested modules?

Commit:
ci: use devcontainer for github actions

Commit:
fix: +x entrypoint.sh

Commit:
fix: cargo install in entrypoint

carlsverre commented on docker/Dockerfile:
```diff
@@ -2,6 +2,13 @@
 RUN curl -L https://github.com/bytecodealliance/wasmtime/releases/download/v${WASMTIME_VERSION}/wasmtime-v${WASMTIME_VERSION}-x86_64-linux.tar.xz \
   | tar -xJ --wildcards --no-anchored --strip-components 1 -C /usr/bin wasmtime
+
+COPY entrypoint.sh /entrypoint.sh
+ENTRYPOINT ["/entrypoint.sh"]
```
It's weird to me to put CI/CD in a script that's setup as the entrypoint rather than putting the commands in the ci/cd pipeline (i.e. in action.yaml or main.yaml).  It's fine to write a script for the commands to run but just weird to have it registered as an entrypoint.  I don't see how this script aligns with all uses of the docker container.

It also seems super weird since all cases of running the Dockerfile already includes access to the source repo and thus the script... so just run the script?

I must be missing something

ricochet replied:
Absolutely, this can be cleaned up and is goofy.

This is from a tutorial I followed: https://bradyjoslin.com/blog/dev-action-container/

I wanted to prove to myself there's nothing special in a devcontainer docker image that would prevent it from being used in an action. Now that we have a passing CI, I'll definitely fix this up.

Commit:
fix: remove entrypoint, change to composite

https://docs.github.com/en/actions/creating-actions/creating-a-composite-action

carlsverre reviewed (approved):
:+1:
